crates/wasi-threads/src/lib.rs

```diff
@@ -0,0 +1,134 @@
+//!
+//! [`wasi-threads`]: https://github.com/WebAssembly/wasi-threads
+
+use anyhow::{anyhow, Result};
+use rand::Rng;
+use std::thread;
+use wasmtime::{Caller, Linker, Module, SharedMemory, Store};
+use wasmtime_wasi::maybe_exit_on_error;
+
+// This name is a function export designated by the wasi-threads specification:
+// https://github.com/WebAssembly/wasi-threads/#detailed-design-discussion
+const WASI_ENTRY_POINT: &str = "wasi_thread_start";
+
+pub struct WasiThreadsCtx<T> {
+    module: Module,
+    linker: Linker<T>,
+}
+
+impl<T: Clone + Send + 'static> WasiThreadsCtx<T> {
+    pub fn new(module: Module, linker: Linker<T>) -> Self {
+        Self { module, linker }
+    }
+
+    pub fn spawn(&self, host: T, thread_start_arg: i32) -> Result<i32> {
+        let module = self.module.clone();
+        let linker = self.linker.clone();
+
+        // Start a Rust thread running a new instance of the current module.
+        let wasi_thread_id = random_thread_id();
+        let builder = thread::Builder::new().name(format!("wasi-thread-{}", wasi_thread_id));
+        builder.spawn(move || {
+            // Convenience function for printing failures, since the `Thread`
+            // has no way to report a failure to the outer context.
+            let fail = |msg: String| {
+                format!(
+                    "wasi-thread-{} exited unsuccessfully: {}",
+                    wasi_thread_id, msg
+                )
+            };
+
+            // Each new instance is created in its own store.
+            let mut store = Store::new(&module.engine(), host);
+            let instance = linker
+                .instantiate(&mut store, &module)
+                .expect(&fail("failed to instantiate".into()));
+            let thread_entry_point = instance
+                .get_typed_func::<(i32, i32), ()>(&mut store, WASI_ENTRY_POINT)
+                .expect(&fail(format!(
+                    "failed to find wasi-threads entry point function: {}",
+                    WASI_ENTRY_POINT
+                )));
+
+            // Start the thread's entry point; any failures are simply printed
+            // before exiting the thread. It may be necessary to handle failures
+            // here somehow, e.g., so that `pthread_join` can be notified if the
+            // user function traps for some reason (TODO).
+            log::trace!(
+                "spawned thread id = {}; calling start function `{}` with: {}",
+                wasi_thread_id,
+                WASI_ENTRY_POINT,
+                thread_start_arg
+            );
+            match thread_entry_point.call(&mut store, (wasi_thread_id, thread_start_arg)) {
+                Ok(_) => {}
+                Err(e) => {
+                    log::trace!("exiting thread id = {} due to error", wasi_thread_id);
+                    let e = maybe_exit_on_error(e);
+                    eprintln!("Error: {:?}", e);
+                    std::process::exit(1);
```

Comment:
One possibility that isn't caught by this is the `.call(..)` function panicking, such as trying to use wasi-nn in a threaded context. If a panic happens that leaves all other threads alive and running and just this thread exits. For that it may be worthwhile, for now at least, to catch panics here as well and abort the whole process if one is caught.